lib/src/scanner.dart

```diff
@@ -288,8 +295,11 @@ class Scanner {
   }
 
   /// Creates a scanner that scans [source].
-  Scanner(String source, {Uri? sourceUrl})
-      : _scanner = SpanScanner.eager(source, sourceUrl: sourceUrl);
+  Scanner(String source,
+      {Uri? sourceUrl, bool recover = false, ErrorListener? errorListener})
```

Comment:
I don't currently see any value in passing in an `errorListener` if `recovery` is `false`, nor any value in setting `recovery` to `true` if there's no `errorListener`. We might consider having the value of `recovery` be implied by `errorListener != null` (that is, make the error listener the only parameter, with `null` implying no recovery).

Reply:
> any value in setting `recovery` to `true` if there's no `errorListener`

In the server where I'm testing this, we care about recovery but not errors (diagnostics aren't produced here, it's in response to a completion request):

https://github.com/dart-lang/sdk/blob/d55dc9c42a4362307e4deb5c97eea2db43d951e4/pkg/analysis_server/lib/src/services/completion/yaml/yaml_completion_generator.dart#L87-L95

I could create a `NullErrorListener` and we could pass that, although it may seem a bit strange to pass a `NullErrorListener` in order to trigger recovery. What do you think?

Reply:
Yeah, that would be a bit odd. We might still want to add an assert that the error listener is always `null` when `recovery` it `false`, because that combination doesn't really make sense. If `recovery` is `false` (and there's an error in the YAML) then the single exception will always be thrown and there's no value to having a listener.

Reply:
Done!